sensing/pointcloud_preprocessor/src/crop_box_filter/crop_box_filter_nodelet.cpp

```diff
@@ -87,49 +91,16 @@ CropBoxFilterComponent::CropBoxFilterComponent(const rclcpp::NodeOptions & optio
 }
 
 void CropBoxFilterComponent::filter(
-  const PointCloud2ConstPtr & input, [[maybe_unused]] const IndicesPtr & indices,
-  PointCloud2 & output)
+  const PointCloud2ConstPtr & input, const IndicesPtr & indices, PointCloud2 & output)
 {
   boost::mutex::scoped_lock lock(mutex_);
-
-  output.data.resize(input->data.size());
-  Eigen::Vector3f pt(Eigen::Vector3f::Zero());
-  size_t j = 0;
-  const auto data_size = input->data.size();
-  const auto point_step = input->point_step;
-  // If inside the cropbox
-  if (!param_.negative) {
-    for (size_t i = 0; i + point_step < data_size; i += point_step) {
-      memcpy(pt.data(), &input->data[i], sizeof(float) * 3);
-      if (
-        param_.min_z < pt.z() && pt.z() < param_.max_z && param_.min_y < pt.y() &&
-        pt.y() < param_.max_y && param_.min_x < pt.x() && pt.x() < param_.max_x) {
-        memcpy(&output.data[j], &input->data[i], point_step);
-        j += point_step;
-      }
-    }
-    // If outside the cropbox
-  } else {
-    for (size_t i = 0; i + point_step < data_size; i += point_step) {
-      memcpy(pt.data(), &input->data[i], sizeof(float) * 3);
-      if (
-        param_.min_z > pt.z() || pt.z() > param_.max_z || param_.min_y > pt.y() ||
-        pt.y() > param_.max_y || param_.min_x > pt.x() || pt.x() > param_.max_x) {
-        memcpy(&output.data[j], &input->data[i], point_step);
-        j += point_step;
-      }
-    }
-  }
-
-  output.data.resize(j);
-  output.header.frame_id = input->header.frame_id;
-  output.height = input->height;
-  output.fields = input->fields;
-  output.is_bigendian = input->is_bigendian;
-  output.point_step = input->point_step;
-  output.is_dense = input->is_dense;
-  output.width = static_cast<uint32_t>(output.data.size() / output.height / output.point_step);
-  output.row_step = static_cast<uint32_t>(output.data.size() / output.height);
+  pcl::PCLPointCloud2::Ptr pcl_input(new pcl::PCLPointCloud2);
+  pcl_conversions::toPCL(*(input), *(pcl_input));
+  impl_.setInputCloud(pcl_input);
+  impl_.setIndices(indices);
+  pcl::PCLPointCloud2 pcl_output;
+  impl_.filter(pcl_output);
+  pcl_conversions::moveFromPCL(pcl_output, output);
```

Comment:
reference : https://github.com/tier4/AutowareArchitectureProposal.iv/pull/773

Reply:
Oh, Now I understand why not the PCL.
However,at present, we have tried many times and found that the crop_box_filter can not be used before "down_sample" and "ground_filter". I think the reason is that "down_sample" and "ground_filter" nodes behind the crop_box_filter use the PCL. The may be some wrong parameters:
https://github.com/autowarefoundation/autoware.universe/blob/main/sensing/pointcloud_preprocessor/src/crop_box_filter/crop_box_filter_nodelet.cpp#L124-L130
more details at #705

 @yukkysaito  says "The conversion is too heavy since the data size is large." I think  we can use "pcl down sample" before "pcl crop box filter" in the xxxx.launch.py to solve the problem.

At present ,Function availability is more important than performance.

Reply:
>However,at present, we have tried many times and found that the crop_box_filter can not be used before "down_sample" and "ground_filter". I think the reason is that "down_sample" and "ground_filter" nodes behind the crop_box_filter use the PCL. 

Thank you for the reply.
It may have some wrong paramえters. I guess we need to see what is wrong.

>@yukkysaito says "The conversion is too heavy since the data size is large." I think we can use "pcl down sample" before "pcl crop box filter" in the xxxx.launch.py to solve the problem.

Yes. It can solve to use "cropbox filter" after "downsample". But I think it is better to be able to use it even before downsample.
Also, there are three problems with downsampling.
- Many DNN algorithms do not assume downsampling.
- INTENSITY is removed by downsample filter(voxel grid filter).
- Errors are increased in the distance to obstacles.  It is harder to guarantee not to hit an obstacle.


>At present ,Function availability is more important than performance.

I agree with you 👍

Reply:
> Also, there are three problems with downsampling.
> 
> * Many DNN algorithms do not assume downsampling.
> * INTENSITY is removed by downsample filter(voxel grid filter).
> * Errors are increased in the distance to obstacles.  It is harder to guarantee not to hit an obstacle.

At present:
Our DNN directly uses the point cloud published by lidar_driver without any pcl handler.
and compared to pcl_crop_box_filter,DNN consumes more and more computing power and memory.

And in euclidean_cluster and ndt_scan_matcher , INTENSITY is useless.

Reply:
I think the original error about invalid point comes from the fact that current implementation assumes that incoming points have height=1. (i.e., unordered points). Rather than changing the cropbox filter itself, I would suggest to change  [L126](https://github.com/autowarefoundation/autoware.universe/blob/bb96ae7b56085a6054ad1ade970bad399faef35a/sensing/pointcloud_preprocessor/src/crop_box_filter/crop_box_filter_nodelet.cpp#L126)
from 
```
  output.height = input->height;
```
to 
```
  output.height = 1;
```

This is reasonable because the pointcloud would become unordered points once you delete certain number of points after cropbox filtering.

Reply:
Thank u! We have tried this modification method before,and it can not cause errors in subsequent nodes("down_sample" and "ground_filter"). But I'm not sure if other nodes using PCL will have errors.

Reply:
thank you!
@mitsudome-r 
 By viewing relevant materials ，I think height to 1 is useful.
 
 The below code is nice.
 output.width = static_cast<uint32_t>(output.data.size() / output.height / output.point_step);
 output.row_step = static_cast<uint32_t>(output.data.size() / output.height);

 Beacuse the PCL is to heavy, I will do an other PR(change the height to 1) to fix the issue.